src/utils/__tests__/internalLinks-test.js

```diff
@@ -37,6 +37,100 @@ describe('isInternalLink', () => {
     [false, 'same domain, double slash', 'https://example.com//#narrow/stream/jest'],
     [false, 'same domain, triple slash', 'https://example.com///#narrow/stream/jest'],
 
+    // These examples may be odd-looking URLs, but are nevertheless valid.
+    [true, 'scheme-relative', '//example.com/#narrow/stream/jest'],
+    [true, 'path-relative, nop path', '.#narrow/stream/jest'],
+    [true, 'path-relative, nop path', './#narrow/stream/jest'],
+    [true, 'path-relative, nop path', '..#narrow/stream/jest'],
+    [true, 'path-relative, nop path', '../#narrow/stream/jest'],
+    [true, 'path-relative, nop path', 'foo/..#narrow/stream/jest'],
+    [true, 'path-relative, nop path', 'foo/../#narrow/stream/jest'],
+    [true, 'path-relative, nop path', 'foo/bar/../../../baz/.././#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/.#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/./#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/..#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/../#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/foo/..#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/foo/../#narrow/stream/jest'],
+    [true, 'path-absolute, nop path', '/foo/bar/../../../baz/.././#narrow/stream/jest'],
```

Comment:
Following all the `../`'s—this would put you up a level _above_ `/`, right, but `new URL` just sets the pathname to `/` instead of throwing an error or something. Odd.

Reply:
Yep -- once the path is at the root, additional `..` path components have no effect. You can see this in a browser: visit e.g.  https://github.com/../../../../zulip/zulip-mobile/pull/4350 . In Chrome for me, the location bar immediately shows the version with the `..` components canonicalized out, before it's even had time to load anything from the server.

In the URL Standard, this is a consequence basically of the "If path is empty" case in the definition of ["shorten a url's path"](https://url.spec.whatwg.org/#shorten-a-urls-path), which is the effect of a "double-dot path segment" [in path state](https://url.spec.whatwg.org/#path-state) in the parser.

It mirrors the behavior of Unix filesystem paths, which URL paths are modeled on / inspired by: for any other directory `..` points to its parent, but for the root directory `/` there is no parent and it points to itself, just like `.`.

Reply:
> It mirrors the behavior of Unix filesystem paths, which URL paths are modeled on / inspired by: for any other directory `..` points to its parent, but for the root directory `/` there is no parent and it points to itself, just like `.`.

Huh, interesting—I didn't know that.